### PR TITLE
Issue #1503: Blog post on tools/strategies for resolving merge conflicts

### PR DESCRIPTION
**Pull request by dmurphy1**
You may just want to pull and review this locally. Depending on when we deploy this, the date may need to be updated in the post and in the file name.

### CODE REVIEW

File: _posts/2016-06-15-git-merge-conflict-strategies.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
+---
+layout: post
+title: "Tools and strategies for resolving Git merge conflicts"
```
Comment (chrisarusso) [inline _posts/2016-06-15-git-merge-conflict-strategies.md]: #nitpicky - I'd prefer 

> Strategies and tools for resolving Git merge conflicts

Strategies seems like the main thing, and is in the URL. Take it or leave it.

File: _posts/2016-06-15-git-merge-conflict-strategies.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+author: Dan Murphy
+tags: git
+summary: An overview of tools and strategies I like to use when I encounter conflicts during merges.
+featured_image: "/assets/img/blog/git_merge_conflict_screenshot.png"
+featured_image_alt: "screenshot of git merge output"
+---
+
+If you use Git for version control and you collaborate frequently, then you are bound to run into merge conflicts from time to time.
+
+In an ideal world you merge upstream updates into your work frequently, and the conflicts are relatively minor and straightforward to resolve.
```
Comment (chrisarusso) [inline _posts/2016-06-15-git-merge-conflict-strategies.md]: maybe another half-sentence to clarify what "upstream updates" mean. Something along the lines of "changes that make their way to production with your parallel feature work" ... or something like that.

File: _posts/2016-06-15-git-merge-conflict-strategies.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+However, from time to time branches diverge significantly and merging them back together takes time and a solid understanding of why each branch may have modified the same hunks of code.
+
+In this post I'm going to review the tools and strategies I like to use when resolving complicated merge conflicts.
+
+## What is merging?
+
+There are many good online tutorials covering the basics of Git merging, so I'll just cover the basics. If you're already fluent in Git, feel free to skip to the strategies and tools section.
```
Comment (chrisarusso) [inline _posts/2016-06-15-git-merge-conflict-strategies.md]: basics twice here reads a bit awkwardly

Comment (chrisarusso) [inline _posts/2016-06-15-git-merge-conflict-strategies.md]: Link to strategies and tools section

File: _posts/2016-06-15-git-merge-conflict-strategies.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+As [Atlassian's branching tutorial](https://www.atlassian.com/git/tutorials/using-branches/git-branch) explains "Merging is Git's way of putting a forked history back together again."
+
+A Git merge is run via the command `git merge <target_branch>`. When merging, Git always merges into the current branch. The branch being merged (aka the target branch) will not be affected.
+
+There are two types of merges: fast-forward merges and 3-way merges.
+
+Fast forward merges occur when the tip of the current branch is a direct ancestor of the target branch. In this case, instead of actually merging the branches, Git simply moves (or fast forwards) the current branch tip up to the target branch tip.
+
+However, a fast-forward merge is not possible if the branches have diverged. In those cases, a 3-way merges is required, which uses a dedicated commit to bring together the two histories.
```
Comment (chrisarusso) [inline _posts/2016-06-15-git-merge-conflict-strategies.md]: > a 3-way merges

File: _posts/2016-06-15-git-merge-conflict-strategies.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+- `git add` on a conflicted file(s) tells Git the conflicts are resolved.
+- `git commit` generates the merge commit completing the merge.
+
+## Strategies and tools
+
+First of all, you should always make sure your working directory is clean before merging. If you have work in progress, either commit it to a temporary branch or stash it. This allows you to easily abandon your merge at any point without losing any of your work. You can abort an in-progress merge using the `git merge --abort` command.
+
+If the merge results in conflicts, I start by finding the merge base using the `git merge-base <tareget_branch> <current_branch>` command. As mentioned previously, the merge base is essentially the commit before the two branches forked.
+
+I then inspect the development history (since the branches diverged) for each file with merge conflicts. This helps me better understand (or simply remember!) each branches intent when modifying the file.
```
Comment (chrisarusso) [inline _posts/2016-06-15-git-merge-conflict-strategies.md]: each branches -> each branch's

**Comment by dmurphy1** [comment]
Incorporated your feedback @chrisarusso

File: _posts/2016-06-15-git-merge-conflict-strategies.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+However, a fast-forward merge is not possible if the branches have diverged. In those cases, a 3-way merge is required, which uses a dedicated commit to bring together the two histories.
+
+In a three way merge, Git's merge algorithm uses the current branch, the target branch, and their merge base in order to incorporate changes from the target branch into the current branch. The merge base is essentially the most recent common ancestor (often this is the commit before the two branches forked). Git uses the merge base to identify what changed since the branches diverged.
+
+Git is smart, and non-overlapping changes (areas of code modified by one branch but not that other) are automatically incorporated into the final merged result. However, sometimes both branches change the same part of the same file, resulting in merge conflicts.
+
+## Merge conflicts
+
+If your branch and the target branch both changed the same part of the same file, then Git won’t be able to merge them cleanly. Each of these discrepancies result in a merge conflict.
```
Comment (AnneTee) [inline _posts/2016-06-15-git-merge-conflict-strategies.md]: This is nitpicky, but "result" should be "results" (Each results in a merge conflict)

**Comment by timstallmann** [comment]
@dmurphy1 this looks great! My only comment is that I would love to see screenshots of the outputs of the different commands for showing file/commit history that you discuss..

**Pull request merged**